pandas/_libs/tslibs/src/datetime/pd_datetime.c

```diff
@@ -0,0 +1,2356 @@
+  char *result = PyObject_Malloc(71);
+  if (result == NULL) {
+    PyErr_NoMemory();
+    return NULL;
+  }
+
+  ret_code = make_iso_8601_timedelta(&tds, result, len);
+  if (ret_code == -1) {
+    PyErr_SetString(PyExc_ValueError,
+                    "Could not convert timedelta value to string");
+    PyObject_Free(result);
+    return NULL;
+  }
+
+  return result;
+}
+
+/*
+ * This function returns a pointer to the DateTimeMetaData
+ * contained within the provided datetime dtype.
+ *
+ * Copied near-verbatim from numpy/core/src/multiarray/datetime.c
+ */
+static PyArray_DatetimeMetaData
+get_datetime_metadata_from_dtype(PyArray_Descr *dtype) {
+  return (((PyArray_DatetimeDTypeMetaData *)dtype->c_metadata)->meta);
+}
+
+static void pandas_datetime_destructor(PyObject *op) {
+  void *ptr = PyCapsule_GetPointer(op, PandasDateTime_CAPSULE_NAME);
+  PyMem_Free(ptr);
+}
+
+static int pandas_datetime_exec(PyObject *module) {
+  PyDateTime_IMPORT;
+  PandasDateTime_CAPI *capi = PyMem_Malloc(sizeof(PandasDateTime_CAPI));
+  if (capi == NULL) {
+    PyErr_NoMemory();
+    return -1;
+  }
+  capi->npy_datetimestruct_to_datetime = npy_datetimestruct_to_datetime;
+  capi->scaleNanosecToUnit = scaleNanosecToUnit;
+  capi->int64ToIso = int64ToIso;
+  capi->NpyDateTimeToEpoch = NpyDateTimeToEpoch;
+  capi->PyDateTimeToIso = PyDateTimeToIso;
+  capi->PyDateTimeToEpoch = PyDateTimeToEpoch;
+  capi->int64ToIsoDuration = int64ToIsoDuration;
+  capi->pandas_datetime_to_datetimestruct = pandas_datetime_to_datetimestruct;
+  capi->pandas_timedelta_to_timedeltastruct =
+      pandas_timedelta_to_timedeltastruct;
+  capi->convert_pydatetime_to_datetimestruct =
+      convert_pydatetime_to_datetimestruct;
+  capi->cmp_npy_datetimestruct = cmp_npy_datetimestruct;
+  capi->get_datetime_metadata_from_dtype = get_datetime_metadata_from_dtype;
+  capi->parse_iso_8601_datetime = parse_iso_8601_datetime;
+  capi->get_datetime_iso_8601_strlen = get_datetime_iso_8601_strlen;
+  capi->make_iso_8601_datetime = make_iso_8601_datetime;
+  capi->make_iso_8601_timedelta = make_iso_8601_timedelta;
+
+  PyObject *capsule = PyCapsule_New(capi, PandasDateTime_CAPSULE_NAME,
+                                    pandas_datetime_destructor);
+  if (capsule == NULL) {
+    PyMem_Free(capi);
+    return -1;
+  }
+
+  // Monkeypatch the top level pandas module to have an attribute for the
+  // C-API. This is required because Python capsules do not support setting
+  // this attribute on anything but the top level package. Ideally not
+  // done when cpython gh-6898 gets implemented
+  PyObject *pandas = PyImport_ImportModule("pandas");
+  if (!pandas) {
+    PyErr_SetString(PyExc_ImportError,
+                    "pd_datetime.c could not import module pandas");
+    Py_DECREF(capsule);
+    // cpython doesn't PyMem_Free(capi) here - mistake or intentional?
+    return -1;
+  }
+
+  if (PyModule_AddObject(pandas, "_pandas_datetime_CAPI", capsule) < 0) {
```

Comment:
Note to self - CPython documentation prefers `PyModule_AddObjectRef` over this. Maybe worth trying instead of using datetime.h pattern:

https://docs.python.org/3/c-api/module.html#c.PyModule_AddObject